helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/AssignmentMetadataStore.java

```diff
@@ -98,7 +108,12 @@ public Map<String, ResourceAssignment> getBestPossibleAssignment() {
         }
       }
     }
-    return _bestPossibleAssignment;
+    // Return defensive copy so that the in-memory assignment is not modified by callers
+    Map<String, ResourceAssignment> result = new HashMap<>(_bestPossibleAssignment);
+    for (Map.Entry<String, ResourceAssignment> entry : _bestPossibleAssignment.entrySet()) {
+      result.put(entry.getKey(), new ResourceAssignment(entry.getValue().getRecord()));
+    }
```

Comment:
_bestPossibleAssignment.size()?

Reply:
And this logic I have seen for multiple lines. Please make them in a function to be called in the place similar.